servers/src/main/java/tachyon/master/file/FileSystemMaster.java

```diff
@@ -1750,6 +1777,63 @@ private void setStateFromEntry(SetStateEntry entry) throws FileDoesNotExistExcep
   }
 
   /**
+   * Checks user's permission on a path
+   *
+   * @param action requested {@link FileSystemAction} by user
+   * @param path the path to check permission on
+   * @param checkParent indicates whether to check the path or its parent
+   * @throws AccessControlException if permission checking fails
+   * @throws InvalidPathException if the path is invalid
+   */
+  private void checkPermission(FileSystemAction action, TachyonURI path,
+      boolean checkParent) throws AccessControlException, InvalidPathException {
+    // bypasses permission checking if security is not enabled.
+    if (MasterContext.getConf().get(Constants.SECURITY_AUTHENTICATION_TYPE).equals("NOSASL")
+        || MasterContext.getConf().get(Constants.SECURITY_AUTHORIZATION_PERMISSION_ENABLED)
+        .equals("false")) {
+      return;
+    }
+
+    // collects inodes info on the path
+    List<Inode> inodes = mInodeTree.collectInodes(path);
+    List<FileInfo> fileInfos = new ArrayList<FileInfo>();
+    for (Inode inode : inodes) {
+      fileInfos.add(inode.generateClientFileInfo(mInodeTree.getPath(inode).toString()));
+    }
+
+    String[] pathComponents = PathUtils.getPathComponents(path.getPath());
+    if (pathComponents.length < fileInfos.size()) {
+      throw new InvalidPathException(ExceptionMessage.PATH_INVALID.getMessage(path.getPath()));
+    }
+
+    // collects user and groups
+    String user = PlainSaslServer.AuthorizedClientUser.get().getName();
+    List<String> groups;
+    try {
+      groups = mGroupMappingService.getGroups(user);
+    } catch (IOException e) {
+      throw new AccessControlException(
+          ExceptionMessage.PERMISSION_DENIED.getMessage(e.getMessage()));
+    }
+
+    // perform permission check
+    if (checkParent) {
+      if ((fileInfos.size() == 1 && pathComponents.length > 1)
+          || (fileInfos.size() == 2 && pathComponents.length == 2)
+          && action.equals(FileSystemAction.WRITE)) {
+        // When checking Write permission on the parent or ancestor of a path, which is root "/",
```

Comment:
Trying to handle a special case when the path is a level under root. I will update the comment for more clarity.